// restmatcher/lib/restmatcher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.supertestWithReqMiddleware = exports.parseCookies = void 0;
const responseMatcher_1 = require("./responseMatcher");
expect.extend({
    statusCode: responseMatcher_1.statusCode,
    toHaveValidationErrorsExactly: responseMatcher_1.toHaveValidationErrorsExactly,
    toHaveAtLeastValidationErrors: responseMatcher_1.toHaveAtLeastValidationErrors,
    toHaveNoValidationErrors: responseMatcher_1.toHaveNoValidationErrors,
    toHaveAnyValidationErrors: responseMatcher_1.toHaveAnyValidationErrors
});
var parseCookies_1 = require("./parseCookies");
Object.defineProperty(exports, "parseCookies", { enumerable: true, get: function () { return parseCookies_1.parseCookies; } });
var supertestWithReqMiddleware_1 = require("./supertestWithReqMiddleware");
Object.defineProperty(exports, "supertestWithReqMiddleware", { enumerable: true, get: function () { return supertestWithReqMiddleware_1.supertestWithReqMiddleware; } });
